File: C#/Word Search/WordSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using _Helper;

public class WordSearch
{
    public Grid<char> grid;

    public WordSearch(string grid)
    {
        string[] rows = grid.Split('\n');

        char[][] gridArray = new char[rows.Length][];

        for (int i = 0; i < rows.Length; i++)
        {
            gridArray[i] = rows[i].ToCharArray();
        }

        this.grid = _Helper.Grid<char>.FromData(gridArray);
    }

    public Dictionary<string, ((int, int), (int, int))?> Search(string[] wordsToSearchFor)
    {
        var results = new Dictionary<string, ((int, int), (int, int))?>();
        foreach (var language in wordsToSearchFor)
        {
            results.Add(language.ToLower(), null);
        }

        // Search each letter in the grid for each language in the list
        this.grid.ForEach((value, row, col) =>
        {
            foreach (var word in wordsToSearchFor)
            {
                var language = word.ToLower();
                
                // All directions longer or equal to word
                var lines = this.grid.CardinalsAndOrdinals(row, col).Where(x => x.Values.Length >= language.Length).ToArray();

                foreach (var line in lines)
                {
                    var potential = string.Join("", line.Values).Substring(0, language.Length);

                    if(language == potential)
                    {
                        var offset = Offsets.OffsetDictionary[line.Direction];

                        int endRow = row + (offset.X * (language.Length - 1));
                        int endCol = col + (offset.Y * (language.Length - 1));

                        // C# exercism solution wants the inverse row - column co-ordinates
                        results[language] = ((col + 1, row + 1), (endCol + 1, endRow + 1));

                    }
                }
            }
        });

        return results;
    }
}

File: JavaScript/word-search.js
const imports = require("./helper/grid");

class WordSearch {
    constructor(input) {
        this.grid = imports.Grid.fromData(input.map(x => x.split('')));
    }
    find(words) {
        const results = {};

        // Search each letter in the grid for each language in the list
        this.grid.forEach((value, row, col) => {
            words.forEach(language => {
                language = language.toLowerCase();

                // All directions longer or equal to word
                let lines = this.grid.cardinalsAndOrdinals(row, col).filter(x => x.values.length >= language.length);
                lines.forEach(line => {
                    let potential = line.values.join('').slice(0, language.length);
                
                    if (language === potential) {
                        let offset = imports.Offsets[line.direction];
                        let endRow = row + (offset.x * (language.length - 1));
                        let endCol = col + (offset.y * (language.length - 1));

                        results[language] = {
                            start: [row + 1, col + 1],
                            end: [endRow + 1, endCol + 1]
                        };

                        return;
                    }
                });
            });
        });

        // Sort the keys of the object based on the order array
        const sortedKeys = Object.keys(results).sort((a, b) => {
            return words.indexOf(a) - words.indexOf(b);
        });

        // Create a new object with sorted keys
        const sortedData = {};
        sortedKeys.forEach(key => {
            sortedData[key] = results[key];
        });

        return sortedData;
    }

    toString() {
        return this.grid.toString();
    }
}

const input = [
    'jefblpepre',
    'camdcimgtc',
    'oivokprjsm',
    'pbwasqroua',
    'rixilelhrs',
    'wolcqlirpc',
    'screeaumgr',
    'alxhpburyi',
    'jalaycalmp',
    'clojurermt',
];

let ws = new WordSearch(input);
let languages = ['clojure', 'elixir', 'ecmascript', 'rust', 'java'];

console.log(ws.toString());
console.log(ws.find(languages));